extern/boostd-data/couchbase/db.go

```diff
@@ -553,6 +554,100 @@ func (db *DB) AddIndexRecords(ctx context.Context, pieceCid cid.Cid, recs []mode
 	return nil
 }
 
+// RemoveAllMetadataForPieceCid
+func (db *DB) RemoveMetadata(ctx context.Context, pieceCid cid.Cid) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_metadata")
+	defer span.End()
+
+	k := toCouchKey(pieceCid.String())
+	metadata, err := db.GetPieceCidToMetadata(ctx, pieceCid)
+	if err != nil {
+		return fmt.Errorf("getting piece %s metadata: %w", pieceCid, err)
+	}
+
+	// Remove all multihashes before, as without Metadata, they are useless
+	// This order is important as metadata.Cursor is required in case RemoveAllRecords fails
+	// and needs to be run manually
+	if err = db.RemoveIndexes(ctx, pieceCid, metadata.BlockCount); err != nil {
+		return fmt.Errorf("failed removing index for piece %s: %w", pieceCid, err)
+	}
+
+	_, err = db.pcidToMeta.Remove(k, &gocb.RemoveOptions{Context: ctx})
+	if err != nil {
+		return fmt.Errorf("removing piece %s metadata: %w", pieceCid, err)
+	}
+
+	return nil
+}
+
+// RemoveAllRecords
+func (db *DB) RemoveIndexes(ctx context.Context, pieceCid cid.Cid, recordCount int) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_indexes")
+	defer span.End()
+
+	// Get the number of shards
+	_, totalShards := getShardPrefixBitCount(recordCount)
+	for i := 0; i < totalShards; i++ {
+		// Get the map of multihash -> offset/size for the shard
+		shardPrefix, err := getShardPrefix(i)
+		if err != nil {
+			return err
+		}
+		cbKey := toCouchKey(pieceCid.String() + shardPrefix)
+		cbMap := db.pieceOffsets.Map(cbKey)
+		recMap, err := cbMap.Iterator()
+		if err != nil {
+			if isNotFoundErr(err) {
+				// If there are no records in a particular shard just skip the shard
+				continue
+			}
+			return fmt.Errorf("getting all records for piece %s: %w", pieceCid, err)
+		}
+
+		// Get each value in the map
+		for mhStr := range recMap {
+			cbKey := toCouchKey(mhStr)
+			getResult, err := db.mhToPieces.Get(cbKey, &gocb.GetOptions{Context: ctx})
+			if err != nil {
+				return nil
+			}
+			var cidStrs []string
+			err = getResult.Content(&cidStrs)
+			if err != nil {
+				return err
+			}
+			if len(cidStrs) <= 1 {
+				_, err = db.mhToPieces.Remove(mhStr, &gocb.RemoveOptions{Context: ctx})
+				if err != nil {
+					return err
+				}
+			}
```

Comment:
I think we should still check to make sure that the piece cid in the array matches the piece cid we're trying to remove. So basically just remove these lines (as we do that check below)
```suggestion
```